src/client/async/async.js

```diff
@@ -0,0 +1,183 @@
+    var uid = evt.data.substring(spf.async.PREFIX_.length);
+    spf.async.run_(uid);
+  }
+};
+
+
+/**
+ * Executes a previously deferred function.
+ *
+ * @param {string|number} uid The UID associated with the function.
+ * @private
+ */
+spf.async.run_ = function(uid) {
+  var fn = spf.async.defers_[uid];
+  if (fn) {
+    delete spf.async.defers_[uid];
+    fn();
+  }
+};
+
+
+/**
+ * Adds a function as a listener for message events.
+ *
+ * @param {!Function} fn The function to add as a listener.
+ * @private
+ */
+spf.async.addListener_ = function(fn) {
+  if (window.addEventListener) {
+    window.addEventListener('message', fn, false);
+  } else if (window.attachEvent) {
+    window.attachEvent('onmessage', fn);
+  }
+};
+
+
+/**
+ * Removes a function as a listener for message events.
+ *
+ * @param {!Function} fn The function to remove as a listener.
+ * @private
+ */
+spf.async.removeListener_ = function(fn) {
+  if (window.removeEventListener) {
+    window.removeEventListener('message', fn, false);
+  } else if (window.detachEvent) {
+    window.detachEvent('onmessage', fn);
+  }
+};
+
+
+/**
+ * Whether the browser supports asynchronous postMessage calls.
+ *
+ * @private {boolean}
+ */
+spf.async.POSTMESSAGE_SUPPORTED_ = (function() {
+  if (!window.postMessage) {
+    return false;
+  }
+  // Use postMessage where available.  But, ensure that postMessage is
+  // asynchronous; the implementation in IE8 is synchronous, which defeats
+  // the purpose.  To detect this, use a temporary "onmessage" listener.
+  var supported = true;
+  var listener = function() { supported = false; };
+  // Add the listener, dispatch a message event, and remove the listener.
+  spf.async.addListener_(listener);
+  window.postMessage('', '*');
+  spf.async.removeListener_(listener);
+  // Return the status.  If the postMessage implementation is correctly
+  // asynchronous, then the value of the `supported` variable will be
+  // true, but if the postMessage implementation is synchronous, the
+  // temporary listener will have executed and set the `supported`
+  // variable to false.
+  return supported;
+})();
+
+
+/**
+ * The prefix to use for message event data to avoid conflicts.
+ *
+ * @private {string}
+ */
+spf.async.PREFIX_ = 'spf:';
+
+
+/**
+ * Map of deferred function calls.
+ * @private {!Object.<!Function>}
+ */
+spf.async.defers_ = {};
+
+
+// Automatic initialization for spf.async.defers_.
+// When built for the bootloader, unconditionally set in state.
+if (SPF_BOOTLOADER) {
+  spf.state.set(spf.state.Key.ASYNC_DEFERS, spf.async.defers_);
+} else {
+  if (!spf.state.has(spf.state.Key.ASYNC_DEFERS)) {
+    spf.state.set(spf.state.Key.ASYNC_DEFERS, spf.async.defers_);
+  }
+  spf.async.defers_ = /** @type {!Object.<!Function>} */ (
+      spf.state.get(spf.state.Key.ASYNC_DEFERS));
+}
+
+// Automatic initialization for spf.state.Key.ASYNC_LISTENER.
+// When built for the bootloader, unconditionally set in state.
+if (SPF_BOOTLOADER) {
```

Comment:
I thought it might make the bootloader optimizations less clear, but I'm going to follow this with a second change specifically for that.